### PR TITLE
executor: skip constraint check when adding record for insert on dup

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
Reduce the BatchGet count if `insert on dup` is the only statements in the transaction.

### What is changed and how it works?
Skip the constraint check when the row is checked.

### Check List 

Tests 

 - Integration test
http://172.16.30.77:40054/dashboard/db/samsung-poc-half-tidb?orgId=1&from=1547026284300&to=1547026584301
image

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8915)

### CODE REVIEW

File: executor/insert_common.go [hunk 1]
```diff
@@ -533,9 +533,8 @@ func (e *InsertValues) handleWarning(err error, logInfo string) {
 
 // batchCheckAndInsert checks rows with duplicate errors.
 // All duplicate rows will be ignored and appended as duplicate warnings.
-func (e *InsertValues) batchCheckAndInsert(rows [][]types.Datum, addRecord func(row []types.Datum) (int64, error)) error {
-	// all the rows will be checked, so it is safe to set BatchCheck = true
-	e.ctx.GetSessionVars().StmtCtx.BatchCheck = true
+func (e *InsertValues) batchCheckAndInsert(rows [][]types.Datum, addRecord func(row []types.Datum,
+	opts ...*table.AddRecordOpt) (int64, error)) error {
```
Comment (lysu) [inline executor/insert_common.go]: it seems this `opts` is no used

File: executor/insert_common.go [hunk 2]
```diff
@@ -562,7 +561,13 @@ func (e *InsertValues) batchCheckAndInsert(rows [][]types.Datum, addRecord func(
 		// There may be duplicate keys inside the insert statement.
 		if rows[i] != nil {
 			e.ctx.GetSessionVars().StmtCtx.AddCopiedRows(1)
-			_, err = addRecord(rows[i])
+			// All the rows are checked, so it is safe to skip the constraint check when adding record.
+			_, err = addRecord(rows[i], &table.AddRecordOpt{
+				CreateIdxOpt: table.CreateIdxOpt{
+					SkipHandleCheck: true,
+					SkipCheck:       true,
```
Comment (lysu) [inline executor/insert_common.go]: maybe `SkipUniqueIndexCheck` is better :D at last if we have many skip flag maybe better use one byte xor, but current way is good to me too

File: executor/insert_common.go (hunk 1 above)
Comment (jackysp) [reply executor/insert_common.go]: The ops are on Line 556.

**Comment by jackysp** [comment]
PTAL @lysu

**Comment by jackysp** [comment]
PTAL @tiancaiamao

File: table/tables/index.go [hunk 3]
```diff
@@ -198,8 +198,12 @@ func (c *index) GenIndexKey(sc *stmtctx.StatementContext, indexedValues []types.
 // Create will return the existing entry's handle as the first return value, ErrKeyExists as the second return value.
 func (c *index) Create(ctx sessionctx.Context, rm kv.RetrieverMutator, indexedValues []types.Datum, h int64,
 	opts ...*table.CreateIdxOpt) (int64, error) {
+	opt := &table.CreateIdxOpt{}
```
Comment (tiancaiamao) [inline table/tables/index.go]: How about

```
var skipUniqueIndexCheck bool
if len(opts) != 0 {
    skipUniqueIndexCheck = opts[0].SkipUniqueIndexCheck
}
...
skipCheck := ctx.GetSessionVars().LightningMode || ctx.GetSessionVars().StmtCtx.BatchCheck || skipUniqueIndexCheck
```

The allocation 
```
opt := &table.CreateIdxOpt{}
```
is unnecessary.

Comment (jackysp) [reply table/tables/index.go]: I'm afraid there will be some other options in the future, then we will have to add more standalone variables.

Comment (tiancaiamao) [reply table/tables/index.go]: When more options is necessary in the future, we have to modify here again, right?
So, why not do that when it's really used, and take the better code style now?

Comment (jackysp) [reply table/tables/index.go]: No need to modify here again, just add some flags at https://github.com/pingcap/tidb/blob/e06c87d3c1705dfd3557f77b502519c7445c5e6e/table/index.go#L31 , and use them at the right place.

File: table/tables/index.go [hunk 4]
```diff
@@ -198,8 +198,12 @@ func (c *index) GenIndexKey(sc *stmtctx.StatementContext, indexedValues []types.
...
+	if len(opts) != 0 {
+		opt = opts[0]
```
Comment (tiancaiamao) [inline table/tables/index.go]: > No need to modify here again

Really???

Comment (jackysp) [reply table/tables/index.go]: You could try. The length of opts could be 0 or 1 only.

Comment (tiancaiamao) [reply table/tables/index.go]: > When more options is necessary in the future, we have to modify here again, right?

**Comment by tiancaiamao** [comment]
PTAL @winkyao

File: executor/insert.go [hunk 5]
```diff
@@ -124,7 +125,12 @@ func (e *InsertExec) batchUpdateDupRows(newRows [][]types.Datum) error {
 		// and key-values should be filled back to dupOldRowValues for the further row check,
 		// due to there may be duplicate keys inside the insert statement.
 		if newRows[i] != nil {
-			newHandle, err := e.addRecord(newRows[i])
+			newHandle, err := e.addRecord(newRows[i], &table.AddRecordOpt{
```
Comment (zz-jason) [inline executor/insert.go]: should we update the comments above this code block?

**Comment by jackysp** [comment]
I will close this PR, because https://github.com/pingcap/tidb/pull/9127 is merged.
What do you think? @tiancaiamao @lysu

**Pull request closed without merging**